### PR TITLE
Delegate hip fat binary loading

### PR DESCRIPTION
**Pull request by Iasonaskrpr**
This PR implements the plugin for issue #72 along with the base class for all HSA static luthier tools.  I have performed tests to determine if the plugin does what we need it to do and it looks ok to me. I can attach some the files.

### CODE REVIEW

**Comment by Iasonaskrpr** [comment]
I have not written any documentation for this. If you want me I can make a .md that describes what and why we do with this plugin and what is the purpose. Also I have left some FIXME comments that I want your input on. Thank you

File: include/luthier/HSATooling/LuthierHSATool.h [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  };
+
+  struct HipSurfaceInfo {
+    void *HostHandle{nullptr};
+    const char *DeviceName{nullptr};
+  };
+
+private:
+  /// @brief Tool Executable Loader, used to register and load HIP code objects
+  ToolExecutableLoader *TEL;
```
Comment (matinraayai) [inline include/luthier/HSATooling/LuthierHSATool.h]: Prefer using a reference instead of a pointer here.

File: include/luthier/HSATooling/LuthierHSATool.h [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  inline LuthierHSATool::~LuthierHSATool() {}
+
+namespace luthier {
+/// @brief Base class for all static HSA tools. This class provides the
+/// necessary structs that have to be filled by LoadHIPFATBinaryInfo plugin so
+/// the static tool registers with the Tool Executable Loader. Variables are
+/// annotated to avoid name demangling, and should not be changed by the tool,
+/// as their contents will be changed by the plugin. This is an abstract class
+/// because it is not useful if not inhereted from by a static tool
+class LuthierHSATool {
```
Comment (matinraayai) [inline include/luthier/HSATooling/LuthierHSATool.h]: Should be a template class `template <typename Tool> class LuthierHSATool`.

File: include/luthier/HSATooling/LuthierHSATool.h [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+public:
+  LuthierHSATool(ToolExecutableLoader *TEL) : TEL(TEL) {}
```
Comment (matinraayai) [inline include/luthier/HSATooling/LuthierHSATool.h]: Add `explicit` keyword to constructors taking a single argument to prevent from implicit conversion.

File: include/luthier/HSATooling/LuthierHSATool.h [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  // Delete default constructor, A tool inhereting from this has no reason to
+  // exist without a TEL
+  LuthierHSATool() = delete;
+  virtual ~LuthierHSATool() = 0;
```
Comment (matinraayai) [inline include/luthier/HSATooling/LuthierHSATool.h]: Set the destructor to `default` instead of `0`.

File: src/lib/Plugins/LoadHIPFATBinaryPlugin/LoadHIPFATBinaryInfoPass.cpp [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+  llvm::Triple T(M.getTargetTriple());
+  // Only operate on host
+  if (T.getArch() == llvm::Triple::ArchType::amdgcn)
+    return llvm::PreservedAnalyses::all();
+  llvm::LLVMContext &C = M.getContext();
+  llvm::StringMap<llvm::GlobalVariable *> NameToVar{};
+  LUTHIER_REPORT_FATAL_ON_ERROR(getAnnotatedValues(M, NameToVar));
+
+  /// FIXME: We may need to keep the global annotations
+  auto AnnotationGV = M.getGlobalVariable("llvm.global.annotations");
```
Comment (matinraayai) [inline src/lib/Plugins/LoadHIPFATBinaryPlugin/LoadHIPFATBinaryInfoPass.cpp]: Selectively delete the variables that need changing; Otherwise, leave annotations as is. They're not hurting anybody and might leave some room for future compiler plugin extensions.

File: src/lib/Plugins/LoadHIPFATBinaryPlugin/LoadHIPFATBinaryInfoPass.cpp [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+  /// through the TEL
+  llvm::Function *RFB = M.getFunction("__hipRegisterFatBinary");
+  llvm::Function *RUFB = M.getFunction("__hipUnregisterFatBinary");
+  llvm::Function *RFUN = M.getFunction("__hipRegisterFunction");
+  llvm::Function *RMV = M.getFunction("__hipRegisterManagedVar");
+  llvm::Function *RDV = M.getFunction("__hipRegisterVar");
+  llvm::Function *RTX = M.getFunction("__hipRegisterTexture");
+  llvm::Function *RSF = M.getFunction("__hipRegisterSurface");
+
+  if (RFB) {
```
Comment (matinraayai) [inline src/lib/Plugins/LoadHIPFATBinaryPlugin/LoadHIPFATBinaryInfoPass.cpp]: If you don't have a fat binary you should exit right away.

File: src/lib/Plugins/LoadHIPFATBinaryPlugin/LoadHIPFATBinaryInfoPass.cpp [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+  /// so the function cannot be deleted since it still would have a use
+  LUTHIER_REPORT_FATAL_ON_ERROR(deleteModuleCtor(M));
+  LUTHIER_REPORT_FATAL_ON_ERROR(
+      deleteFunction(M.getFunction("__hip_module_ctor")));
+  LUTHIER_REPORT_FATAL_ON_ERROR(
+      deleteFunction(M.getFunction("__hip_register_globals")));
+  ///  Delete all functions that call __hipRegisterFatBinary and then delete
+  /// __hipRegisterFatBinary as well, we do the same for
+  /// __hipUnregisterFatBinary as well below
+  for (auto *user : RFB->users()) {
```
Comment (matinraayai) [inline src/lib/Plugins/LoadHIPFATBinaryPlugin/LoadHIPFATBinaryInfoPass.cpp]: Camel Case name.

File: include/luthier/HSATooling/LuthierHSATool.h [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  __attribute__((used)) TOOLNAME::HipManagedVarInfo *TOOLNAME::HipManagedVars{ \
+      nullptr};                                                                \
+  __attribute__((used)) unsigned long long TOOLNAME::HipManagedVarsSize{0};    \
+  __attribute__((used)) TOOLNAME::HipTextureInfo *TOOLNAME::HipTextureVars{    \
+      nullptr};                                                                \
+  __attribute__((used)) unsigned long long TOOLNAME::HipTextureVarsSize{0};    \
+  __attribute__((used)) TOOLNAME::HipSurfaceInfo *TOOLNAME::HipSurfaceVars{    \
+      nullptr};                                                                \
+  __attribute__((used)) unsigned long long TOOLNAME::HipSurfaceVarsSize{0};    \
+  inline LuthierHSATool::~LuthierHSATool() {}
```
Comment (matinraayai) [inline include/luthier/HSATooling/LuthierHSATool.h]: Don't need this.

File: src/lib/ToolingCommon/LoadHIPFATBinaryInfoPass.cpp [hunk 9]
```diff
@@ -0,0 +1,526 @@
+//===-- LoadHIPFATBinaryInfo.cpp --------------------------------*- C++ -*-===//
```
Comment (matinraayai) [inline src/lib/ToolingCommon/LoadHIPFATBinaryInfoPass.cpp:1]: Put this under `ToolingCommon`, put the header file under the public `include` directory under `Tooling`. Add this pass to the `OptPlugin` under `Plugins/OptPlugin`. Then write a lit test or two and add them under src/test.

File: src/lib/Plugins/LoadHIPFATBinaryPlugin/LoadHIPFATBinaryInfoPass.cpp [hunk 10]
```diff
@@ -0,0 +1,526 @@
...
+  NewTable->takeName(OldVar);
+  /// RAUW the new variable. We do a bitcast in case the type is different
+  OldVar->replaceAllUsesWith(
+      llvm::ConstantExpr::getBitCast(NewTable, OldVar->getType()));
+
+  OldVar->eraseFromParent();
+  return llvm::Error::success();
+}
+
+/// @brief Same function as the other one, but this is specifically for the fat
```
Comment (matinraayai) [inline src/lib/Plugins/LoadHIPFATBinaryPlugin/LoadHIPFATBinaryInfoPass.cpp]: Use slash doxygen commands.

**Review by matinraayai — CHANGES_REQUESTED** [review]
Please see comments.

File: src/lib/Plugins/LoadHIPFATBinaryPlugin/LoadHIPFATBinaryInfoPass.cpp (hunk 5 above)
Comment (Iasonaskrpr) [reply src/lib/Plugins/LoadHIPFATBinaryPlugin/LoadHIPFATBinaryInfoPass.cpp]: I was wrong here, all annotations stay valid since I RAUW with the new variable and then erase the old one. Also since llvm.global.annotations only holds an opaque ptr to this variable, it doesn't care what that variable's type is. I will not delete any of them, unless you want me to delete the ones that were actually changed.

**Comment by Iasonaskrpr** [comment]
I seperated the plugin, where one file is the registration of the plugin under Plugins only and the other is the pass under ToolingCommon. I could do the same for EmbedIModulePlugin but currently it does not compile and I would either have to fix it or have no idea if the seperation I made was correct or not so I don't think it is a good idea for now. But combining the 2 in one plugin I think would make it cleaner and make it possible to add testing for it. I tested the fatbin plugin and everything works as expected and compiles fine. I also added some tests and if I have time I will also cover some edge cases because now I only test the happy path. Let me know if you need me to change anything!

**Comment by matinraayai** [comment]
@Iasonaskrpr merged. thank you!
